rollbar/resource_notification.go

```diff
@@ -0,0 +1,282 @@
+										Description: "Operation",
+										Type:        schema.TypeString,
+										Optional:    true,
+									},
+									"value": {
+										Description: "Value",
+										Type:        schema.TypeString,
+										Optional:    true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"config": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"users": {
+							Type:        schema.TypeList,
+							Optional:    true,
+							Description: "Users (email)",
+							Elem:        &schema.Schema{Type: schema.TypeString},
+						},
+						"teams": {
+							Type:        schema.TypeList,
+							Optional:    true,
+							Description: "Teams (email)",
+							Elem:        &schema.Schema{Type: schema.TypeString},
+						},
+						"message_template": {
+							Description: "Message template (slack)",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+						"channel": {
+							Description: "Channel (slack)",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+						"show_message_buttons": {
+							Description: "Show message buttons (slack)",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+						"service_key": {
+							Description: "Service key (pagerduty)",
+							Type:        schema.TypeString,
+							Optional:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func find(slice []string, val string) bool {
+	for _, item := range slice {
+		if item == val {
+			return true
+		}
+	}
+	return false
+}
+
+func parseSet(setName string, d *schema.ResourceData) map[string]interface{} {
+	setMap, ok := d.GetOk(setName)
+	properSetMap := map[string]interface{}{}
+
+	if ok {
+		set := setMap.(*schema.Set).List()
+		for _, s := range set {
+			properSetMap, ok = s.(map[string]interface{})
+			if ok {
+				return properSetMap
+			}
+		}
+	}
+	return map[string]interface{}{}
+}
+
+func parseRule(d *schema.ResourceData) (string, interface{}) {
+	rule := parseSet("rule", d)
+
+	var trigger string
+	var filters interface{}
+	for key, value := range rule {
+		if key == "trigger" {
+			trigger = value.(string)
+		} else {
+			filters = value
+		}
+	}
+	return trigger, filters
+}
+
+func cleanConfig(channel string, config map[string]interface{}) map[string]interface{} {
+	returnSetMap := map[string]interface{}{}
+	for key, v := range config {
+		if find(configMap[channel], key) {
+			returnSetMap[key] = v
+		}
+	}
+	return returnSetMap
+}
+
+func resourceNotificationCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+
+	trigger, filters := parseRule(d)
+	channel := d.Get("channel").(string)
+	config := parseSet("config", d)
+	config = cleanConfig(channel, config)
+	l := log.With().Str("channel", channel).Logger()
+
+	l.Info().Msg("Creating rollbar_notification resource")
+
+	c := m.([]*client.RollbarAPIClient)[1]
```

Comment:
i see that this PR is introduing a new kind of client, and are accessing them by index. 

From my understanding, instead of returning a list of clients, we could create a type that would take in either the c client, the pc client, or both. 

This way we can access them by their name as opposed to index, which would heavily confuse people if there's a bug and we're relying on indices. I'm not too familiar with Go, so curious to hear your thoughts on this?

Reply:
it's not really about go, but about the terraform api. providerConfigure should return the client, which is then passed to ConfigureContextFunc: https://github.com/rollbar/terraform-provider-rollbar/blob/97b5f9b0a851a3ccedf2c99a385a748ef4768318/rollbar/provider.go#L75
 this is the limitation of terraform plugin

I could not find a better way to implement that ( I also try the way you decried )